include/deal.II/matrix_free/vector_data_exchange.h

```diff
@@ -247,6 +247,273 @@ namespace internal
+          const ArrayView<const Number> &             locally_owned_array,
+          const std::vector<ArrayView<const Number>> &shared_arrays,
+          const ArrayView<Number> &                   ghost_array,
+          std::vector<MPI_Request> &                  requests) const;
+
+        template <typename Number>
+        void
+        import_from_ghosted_array_start_impl(
+          const VectorOperation::values               vector_operation,
+          const unsigned int                          communication_channel,
+          const ArrayView<const Number> &             locally_owned_array,
+          const std::vector<ArrayView<const Number>> &shared_arrays,
+          const ArrayView<Number> &                   ghost_array,
+          const ArrayView<Number> &                   temporary_storage,
+          std::vector<MPI_Request> &                  requests) const;
+
+        template <typename Number>
+        void
+        import_from_ghosted_array_finish_impl(
+          const VectorOperation::values               vector_operation,
+          const ArrayView<Number> &                   locally_owned_storage,
+          const std::vector<ArrayView<const Number>> &shared_arrays,
+          const ArrayView<Number> &                   ghost_array,
+          const ArrayView<const Number> &             temporary_storage,
+          std::vector<MPI_Request> &                  requests) const;
+
+        template <typename Number>
+        void
+        reset_ghost_values_impl(const ArrayView<Number> &ghost_array) const;
+
+      private:
+        /**
+         * Global communicator.
+         */
+        const MPI_Comm comm;
+
+        /**
+         * Shared-memory sub-communicator.
+         */
+        const MPI_Comm comm_sm;
+
+        /**
+         * Number of locally-owned vector entries.
+         */
+        const unsigned int n_local_elements;
+
+        /**
+         * Number of ghost vector entries.
+         */
+        const unsigned int n_ghost_elements;
+
+        /**
+         * Number of global vector entries.
+         */
+        const types::global_dof_index n_global_elements;
+
+        /**
+         * A variable caching the number of ghost indices in a larger set of
+         * indices by rank.
+         */
+        std::vector<unsigned int> n_ghost_indices_in_larger_set_by_remote_rank;
+
+        /**
+         * The set of indices that appear for an IndexSet that is a subset of a
+         * larger set for each rank in a compressed manner.
+         */
+        std::pair<std::vector<unsigned int>,
+                  std::vector<std::pair<unsigned int, unsigned int>>>
+          ghost_indices_subset_data;
+
+        /**
+         * An array that contains information which processors my ghost indices
+         * belong to, at which offset and how many those indices are
+         */
+        std::vector<
+          std::pair<unsigned int, std::pair<unsigned int, unsigned int>>>
+          ghost_targets_data;
```

Comment:
I'm thinking about the data structure here. Would be lose a lot if we made this an `std::array<unsigned int, 3>` and state that the first index is the processor rank, the second the offset within the ghost array, and the third the number of indices? To me, accessing `ghost_targets_data[1]` would at least not be worse than `ghost_targets_data.second.first`.

Reply:
I have changing this one and import_targets_data.